Provide more detail on errors

petermattis opened the pull request:
When an error occurs exec'ing various processes, include both the output
and the error in the message. For example:

```
~ roachprod start local
local: starting 2/2
0: bash: /Users/pmattis/local/1/cockroach: is a directory
: Process exited with status 126:
1: bash: /Users/pmattis/local/2/cockroach: is a directory
: Process exited with status 126:
2018/03/06 18:43:27 command failed
```

Fixes #125

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/roachprod/126)
<!-- Reviewable:end -->

petermattis commented:
Cc @arjunravinarayan

nvanbenschoten commented on install/cockroach.go:
```diff
@@ -69,7 +70,7 @@ func getCockroachVersion(c *SyncedCluster, i int, host, user string) (*version.V
 	out, err := session.CombinedOutput(cmd)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "%s", out)
```
Can't these be replaced with `errors.Wrap(err, out)`? If so, slight preference everywhere to create strings with `errors.Wrap(err, out).Error()`.

petermattis replied:
`out` is a `[]byte`. The default formatting for `[]byte` is unreadable. I could make this `errors.Wrap(err, string(out))`, though I prefer the explicitness of `errors.Wrapf`.

nvanbenschoten replied:
In that case, I like this approach better too.

benesch commented:
```
0: bash: /Users/pmattis/local/1/cockroach: is a directory
: Process exited with status 126:
1: bash: /Users/pmattis/local/2/cockroach: is a directory
: Process exited with status 126:
2018/03/06 18:43:27 command failed
```

That's not very pretty output. I think printing `err: output` instead of `output: err` would go miles:

```
0: Process exited with status 126: bash: /Users/pmattis/local/1/cockroach: is a directory
1: Process exited with status 126: bash: /Users/pmattis/local/2/cockroach: is a directory
2018/03/06 18:43:27 command failed
```

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[install/cockroach.go, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6xjIlLfN5YsISp91HM-r1-73:-L6ymJi7V4cYDZCuX6-n:b-oyjces) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L73)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In that case, I like this approach better too.
</blockquote></details>

Per my root comment I think this would be better as `fmt.Errorf("%s: %s", err, strings.TrimSpace(out))`. 

---

*[install/cockroach.go, line 324 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6ykTHgQ-netgis1vVG:-L6ykTHgQ-netgis1vVH:b-upz8yg) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L324)):*
> ```Go
> 			binary + " start " + strings.Join(args, " ") +
> 			" >> " + logDir + "/cockroach.stdout 2>> " + logDir + "/cockroach.stderr"
> 		return session.CombinedOutput(cmd)
> ```

Have you run this against a remote cluster where `cockroach start` fails? That's the case Arjun filed #125 about and I think it's the one case you don't handle here! It's also the hardest case; the output of `cockroach start` is swallowed by these redirections.

I think the simplest approach is to tack on make `cmd` look like this:

```
cockroach start --insecure .... > logs/cockroach.stdout 2> logs/cockroach.stderr || cat logs/cockroach.stdout logs/cockroach.stderr 2> /dev/null
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-:-L6ylRZiVnAAK-kKAgIK:bgwd3uo)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[install/cockroach.go, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6xjIlLfN5YsISp91HM-r1-73:-L6ymYiBw9RsmotLMZAU:beqkw2r) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L73)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Per my root comment I think this would be better as `fmt.Errorf("%s: %s", err, strings.TrimSpace(out))`. 
</blockquote></details>

Make that `fmt.Errorf("%s: %s", err, bytes.TrimSpace(out))`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/roachprod/126)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks for doing this!

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[install/cockroach.go, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6xjIlLfN5YsISp91HM-r1-73:-L6zKoxMmUmhXmBpDx4_:bn7r5tv) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L73)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Make that `fmt.Errorf("%s: %s", err, bytes.TrimSpace(out))`.
</blockquote></details>

I think it's worth sticking with `pkg/errors` because those contain [stack traces](https://godoc.org/github.com/pkg/errors#hdr-Retrieving_the_stack_trace_of_an_error_or_wrapper) which I also think we should leverage; it's often annoying to have to grep your way to the source of a particular error, and sometimes it's even impossible.

Any error in `roachtest` is usually fatal, so we might as well `%+v` them, and then the stack trace will go a long way. (Not suggesting that kind of work be done in this PR).

---

*[install/cockroach.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6zLOkMw56X5TSnluoW:-L6zLOkMw56X5TSnluoX:bbek3bk) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L230)):*
> ```Go
> 				cmd += `tar xf -`
> 				if out, err := session.CombinedOutput(cmd); err != nil {
> 					return nil, errors.Wrapf(err, "%s", out)
> ```

For greppability, it would help to do `errors.Wrapf(err, "starting node: %s", out)`.

---

*[install/cockroach.go, line 324 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6ykTHgQ-netgis1vVG:-L6zLl_Tn24dH-OLCeZ2:b-4lf094) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L324)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Have you run this against a remote cluster where `cockroach start` fails? That's the case Arjun filed #125 about and I think it's the one case you don't handle here! It's also the hardest case; the output of `cockroach start` is swallowed by these redirections.

I think the simplest approach is to tack on make `cmd` look like this:

```
cockroach start --insecure .... > logs/cockroach.stdout 2> logs/cockroach.stderr || cat logs/cockroach.stdout logs/cockroach.stderr 2> /dev/null
```
</blockquote></details>

@benesch but that'll change the exit code to success, won't it? (I wrote the `NB:` above and would love to see this fixed as it's about as opaque as it gets). Do we really have to resort to something like below?

```
./cockroach start ... || (x=$?; cat ... && exit $?)
```

---

*[install/cockroach.go, line 355 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6zMDuRH49jiHYa7dM3:-L6zMDuS2cRAvx8XxymY:b-27ies2) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L355)):*
> ```Go
> 			out, err := session.CombinedOutput(cmd)
> 			if err != nil {
> 				msg = fmt.Sprintf("%s: %v", out, err)
> ```

The old variant was a lot more greppable, I'd try to maintain that.

---

*[install/cockroach.go, line 434 at r1](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-L6zMgqKAiLdIH4c20Lx:-L6zMgqKAiLdIH4c20Ly:bdk1mt5) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L434)):*
> ```Go
> 		out, err := session.CombinedOutput(cmd)
> 		if err != nil {
> 			resultChan <- result{node: c.Nodes[i], output: fmt.Sprintf("%s: %v", out, err)}
> ```

Here too it would help to make it greppable. Maybe inserting an `errors.Wrapf(err, "SQL on node %", i)` above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/roachprod/126#-:-L6zMxhjxkUgge58Bu98:bxgavm9)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[install/cockroach.go, line 324 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6ykTHgQ-netgis1vVG:-L7-eLydCC4ElqZGOPaO:b-v6vkae) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L324)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

@benesch but that'll change the exit code to success, won't it? (I wrote the `NB:` above and would love to see this fixed as it's about as opaque as it gets). Do we really have to resort to something like below?

```
./cockroach start ... || (x=$?; cat ... && exit $?)
```
</blockquote></details>

The earlier `getCockroachVersion` call will detect if the binary isn't runnable at all. But you have a good point that I didn't completely solve the problem. For example, `roachprod start local -a "--foo"` will still print a terse message. I've fixed this by removing the stdout and stderr redirection. That was a historical artifact from `roachperf` when I was using `--logtostderr`. Now that I write this, I'm worried that not redirecting stdout/stderr can lead to problems. Hmm, maybe I should take @tschottdorf's approach. 

---

*[install/cockroach.go, line 355 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6zMDuRH49jiHYa7dM3:-L7-fIbJ5dUTpxgO9HON:b-gs690) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L355)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The old variant was a lot more greppable, I'd try to maintain that.
</blockquote></details>

Good point. I've tweaked this to include the command that failed which provides both greppability and additional detail.

---

*[install/cockroach.go, line 434 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6zMgqKAiLdIH4c20Lx:-L7-gbS_Ac4Kt3Ut2qGn:byb7lit) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L434)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Here too it would help to make it greppable. Maybe inserting an `errors.Wrapf(err, "SQL on node %", i)` above.
</blockquote></details>

Similar to above, I've included the command that failed in the message. I've also fixed this so that it actually displays this output

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[install/cockroach.go, line 324 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6ykTHgQ-netgis1vVG:-L7-n4djEmWvZKB1WRgF:b-mo1v3k) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L324)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The earlier `getCockroachVersion` call will detect if the binary isn't runnable at all. But you have a good point that I didn't completely solve the problem. For example, `roachprod start local -a "--foo"` will still print a terse message. I've fixed this by removing the stdout and stderr redirection. That was a historical artifact from `roachperf` when I was using `--logtostderr`. Now that I write this, I'm worried that not redirecting stdout/stderr can lead to problems. Hmm, maybe I should take @tschottdorf's approach. 
</blockquote></details>

Ok, I've added @tschottdorf's nugget and verified it works. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[install/cluster_synced.go, line 382 at r2](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L70AA1mEIUAa5OU_XTG:-L70AA1mEIUAa5OU_XTH:b-938ypt) ([raw file](https://github.com/cockroachdb/roachprod/blob/5f662ff9da41b72604aec91331bc1f62cc137b35/install/cluster_synced.go#L382)):*
> ```Go
> 		var s string
> 		if err != nil {
> 			s = fmt.Sprintf("%s: %v", out, err)
> ```

Did you want to `~ ` this as well?

---

*[install/cockroach.go, line 324 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6ykTHgQ-netgis1vVG:-L709LzD0I27RMwv9AzI:bz4xbps) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L324)):*
> The earlier getCockroachVersion call will detect if the binary isn't runnable at all.

Hmm, is that new? That didn't work for me a few days ago and Arjun's issue also seems to indicate that it didn't work for him yesterday.

I'm not seeing the nugget, did you forget to push?

---

*[install/cockroach.go, line 355 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6zMDuRH49jiHYa7dM3:-L709dVv4tUfps1yEaU0:bdrdjeo) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L355)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Good point. I've tweaked this to include the command that failed which provides both greppability and additional detail.
</blockquote></details>

This should do the trick, but it's not quite greppable yet, right? What would you grep for? The best I can do now is search for "~ " and then figure out which of the lines it is.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[install/cluster_synced.go, line 382 at r2](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L70AA1mEIUAa5OU_XTG:-L70BHvA0_BiDjZ51rte:b-wprzto) ([raw file](https://github.com/cockroachdb/roachprod/blob/5f662ff9da41b72604aec91331bc1f62cc137b35/install/cluster_synced.go#L382)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Did you want to `~ ` this as well?
</blockquote></details>

No, that was intentional to keep the output smaller. This code (`CockroachVersion`) is going to die very soon. See #79.

---

*[install/cockroach.go, line 73 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6xjIlLfN5YsISp91HM-r1-73:-L70BVB2BEk67jaCiRth:bh87o3k) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L73)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think it's worth sticking with `pkg/errors` because those contain [stack traces](https://godoc.org/github.com/pkg/errors#hdr-Retrieving_the_stack_trace_of_an_error_or_wrapper) which I also think we should leverage; it's often annoying to have to grep your way to the source of a particular error, and sometimes it's even impossible.

Any error in `roachtest` is usually fatal, so we might as well `%+v` them, and then the stack trace will go a long way. (Not suggesting that kind of work be done in this PR).
</blockquote></details>

`SyncedCluster.Parallel` formats the error with `%+v` now. The stack traces are usually quite small.

---

*[install/cockroach.go, line 230 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6zLOkMw56X5TSnluoW:-L70BedzD7NSDyEjz3zm:bt3t5z0) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L230)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

For greppability, it would help to do `errors.Wrapf(err, "starting node: %s", out)`.
</blockquote></details>

See other comment about grepping for the command text (i.e. `tar xf`).

---

*[install/cockroach.go, line 324 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6ykTHgQ-netgis1vVG:-L70Ah0S3vYnWvopdkGk:btm86fu) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L324)):*
Yes, I forgot to push. Done.

> Hmm, is that new? That didn't work for me a few days ago and Arjun's issue also seems to indicate that it didn't work for him yesterday.

No, it's not new, but the error reporting is. What Arjun was seeing was an execution failure of `cockroach version`. Prior to this PR nothing interesting was output. Now, `getCockroachVersion()` includes an expanded error message. 

---

*[install/cockroach.go, line 355 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6zMDuRH49jiHYa7dM3:-L70B8476xhRK5SE82qp:b-1n0roy) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L355)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This should do the trick, but it's not quite greppable yet, right? What would you grep for? The best I can do now is search for "~ " and then figure out which of the lines it is.
</blockquote></details>

The error message contains the complete command (e.g. `SET CLUSTER SETTING`) which you can then grep for. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[install/cockroach.go, line 324 at r1](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-L6ykTHgQ-netgis1vVG:-L70G906-6wnEzRSuzhG:b6ibvbl) ([raw file](https://github.com/cockroachdb/roachprod/blob/060b01047b2c25715b9b30748f171659993a51ff/install/cockroach.go#L324)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes, I forgot to push. Done.

> Hmm, is that new? That didn't work for me a few days ago and Arjun's issue also seems to indicate that it didn't work for him yesterday.

No, it's not new, but the error reporting is. What Arjun was seeing was an execution failure of `cockroach version`. Prior to this PR nothing interesting was output. Now, `getCockroachVersion()` includes an expanded error message. 
</blockquote></details>

Ah, TIL. I never even traced the error source to that.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/roachprod/126#-:-L70GEPq2f7sptnju9oj:bnfp4nl)*
<!-- Sent from Reviewable.io -->
